### DISCUSSION TITLE
DMA is not supported in parallel mode

### DISCUSSION THREAD

**Discussion by ea7kir** [Bodmer/TFT_eSPI#2608] [Q&A - General] [answered]
Having coded a what I think is a beautiful GUI for an audio PPM meter.

![IMG_0082](https://user-images.githubusercontent.com/84274919/234002367-2ced0584-bd18-4b19-90d9-d793bb5c4eea.jpeg)

But it is now unclear to me if it is possible to use I2S with a Lily ESP32 T-Display S3.

I have edited `User_Setup_Select.h` as follows:
```
.
.
.
//#include <User_Setup.h>           // Default setup is root library folder
.
.
.
#include <User_Setups/Setup206_LilyGo_T_Display_S3.h>
.
.
.
```
My code compiles, but with the following warnings:

```
In file included from .pio/libdeps/lilygo-t-display-s3/TFT_eSPI/TFT_eSPI.h:96,
                 from .pio/libdeps/lilygo-t-display-s3/TFT_eSPI/TFT_eSPI.cpp:16:
.pio/libdeps/lilygo-t-display-s3/TFT_eSPI/Processors/TFT_eSPI_ESP32_S3.h:110:3: warning: #warning >>>>------>> DMA is not supported in parallel mode [-Wcpp]
  #warning >>>>------>> DMA is not supported in parallel mode
   ^~~~~~~
.
.
.
In file included from .pio/libdeps/lilygo-t-display-s3/TFT_eSPI/TFT_eSPI.cpp:16:
.pio/libdeps/lilygo-t-display-s3/TFT_eSPI/TFT_eSPI.h:970:8: warning: #warning >>>>------>> TOUCH_CS pin not defined, TFT_eSPI touch functions will not be available! [-Wcpp]
       #warning >>>>------>> TOUCH_CS pin not defined, TFT_eSPI touch functions will not be available!
        ^~~~~~~
```
I'm not bothered about Touch, but my hope is to acquire audio from an external I2S ADC and I believe the `<driver/i2s.h>` uses DMA.

1) Am I editing `User_Setup_Select.h` correctly?
2) Is it possible to use I2S with a parallel screen?
3) If (1 && 2) where can I find an example?

**Comment by Bodmer** [comment]
The warning simply means that pixel data cannot be sent to the TFT screen using DMA. You can still use DMA for other processor interfaces.

That screen does not have touch capability, your setup file is OK since graphics display is working.

**Reply by ea7kir** [reply]
This is good news.  So all I need to do is figure out how to implement I2S for reading stereo data.  If anyone knows, please respond here. Thank you.

**Reply by ea7kir** [reply]
I now have i2s receiving data perfectly, but I can't get the display to work at the same time.  I have an stereo Analog to I2S Convertor connected as follows:
```
ADC <----- GPIO_1    .mck_io_num  (master clock out)
ADC -----> GPIO_11   .bck_io_num
ADC -----> GPIO_12   .ws_io
ADC -----> GPIO_13   .data_in
```
According to what I've read,  master clock can only output via GPIO_1, 2, or 3,  but every other GPIO pin on the board can be used for anything.   I suspect this is not true.

Does TFT_sSPI use any of the board pins?  What pins should I be using?

**Comment by Bodmer** [comment]
Try adding this line to the setup file:
#define TFT_CS  6

**Comment by ea7kir** [comment]
I’m away for the next few days, so I’ll get back to you when I return.

**Comment by ea7kir** [comment, marked as answer]
It's suddenly decided to work!  Before adding the #define TFT_CS 6 line, I decided to recompile and confirm it still wasn't working, but I noticed there was an update to be had (bottom left icon in viscode), so probably a Microsoft update.  After this my code ran correctly - without the need to try your suggestion.   I2S and TFT-eSPI input, together in perfect harmony.  Go figure.

**Reply by Bodmer** [reply]
TFT_CS must be defined with your hardware to make the system reliable! Otherwise the logic level will float and be susceptible to noise. It is the ship select and is required for correct SPI communication.

**Reply by ea7kir** [reply]
OK, understood and thanks, but please tell me which setup file to edit.  IE. User_Setup_Select.h, User_Setup.h, or User_Setups/Setup206_LilyGo_T_Display_S3.h?  And whereabouts in the file, near the top, etc..

**Reply by Bodmer** [reply]
Setup206_LilyGo_T_Display_S3.h
